File: src/test_sistema.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <algorithm>
#include <vector>

// Framework minimo de testes
static int testes_passados = 0;
static int testes_falhos   = 0;

void check(bool condicao, const std::string& nome) {
    if (condicao) {
        std::cout << "[OK]     " << nome << "\n";
        testes_passados++;
    } else {
        std::cout << "[FALHOU] " << nome << "\n";
        testes_falhos++;
    }
}

// 1. Encoder: deteccao de transicao 
void test_encoder() {
    std::cout << "- Encoder\n";

    bool ultimo = false;
    int posicao_x = 0;

    // Primeira transicao 0->1
    bool atual = true;
    if (atual != ultimo) { posicao_x++; ultimo = atual; }
    check(posicao_x == 1, "Transicao 0->1 incrementa posicao_x");

    // Mesmo estado: sem incremento
    atual = true;
    if (atual != ultimo) { posicao_x++; ultimo = atual; }
    check(posicao_x == 1, "Estado repetido nao incrementa posicao_x");

    // Segunda transicao 1->0
    atual = false;
    if (atual != ultimo) { posicao_x++; ultimo = atual; }
    check(posicao_x == 2, "Transicao 1->0 incrementa posicao_x");

    // Sequencia de 10 transicoes
    for (int i = 0; i < 10; i++) {
        atual = !ultimo;
        if (atual != ultimo) { posicao_x++; ultimo = atual; }
    }
    check(posicao_x == 12, "10 transicoes consecutivas incrementam 10 vezes");
}

// 2. Filtro de media movel (janela de 2 amostras) 
void test_filtro_media_movel() {
    std::cout << "- Filtro de media movel\n";

    float janela[2] = {};
    int   idx       = 0;
    bool  cheia     = false;

    auto inserir = [&](float v) -> float {
        janela[idx % 2] = v;
        idx++;
        if (idx >= 2) cheia = true;
        int n = cheia ? 2 : idx;
        float soma = 0;
        for (int i = 0; i < n; i++) soma += janela[i];
        return soma / n;
    };

    float m1 = inserir(100.0f);
    check(std::fabs(m1 - 100.0f) < 0.01f, "Primeira amostra retorna valor puro (janela incompleta)");

    float m2 = inserir(200.0f);
    check(std::fabs(m2 - 150.0f) < 0.01f, "Media de 100 e 200 = 150");

    float m3 = inserir(50.0f);
    check(std::fabs(m3 - 125.0f) < 0.01f, "Janela deslizante: media de 200 e 50 = 125");

    float m4 = inserir(50.0f);
    check(std::fabs(m4 - 50.0f) < 0.01f,  "Duas amostras identicas: media = proprio valor");
}

// 3. Calculo do baseline
void test_baseline() {
    std::cout << "- Baseline\n";

    const int BASELINE_AMOSTRAS = 15;
    float baseline      = 0.0f;
    float baseline_soma = 0.0f;
    int   baseline_n    = 0;

    // Amostras validas (> 100)
    for (int i = 0; i < BASELINE_AMOSTRAS; i++) {
        float v = 145.0f;
        if (baseline_n < BASELINE_AMOSTRAS && v > 100.0f) {
            baseline_soma += v;
            baseline_n++;
            if (baseline_n == BASELINE_AMOSTRAS)
                baseline = baseline_soma / BASELINE_AMOSTRAS;
        }
    }
    check(baseline_n == BASELINE_AMOSTRAS, "Acumula exatamente 15 amostras");
    check(std::fabs(baseline - 145.0f) < 0.01f, "Baseline calculado corretamente");

    // Amostras invalidas (< 100) devem ser rejeitadas
    int n_invalido = 0;
    float v_invalido = 50.0f;
    if (v_invalido > 100.0f) n_invalido++;
    check(n_invalido == 0, "Amostras abaixo de 100 sao rejeitadas no startup");

    // Amostra adicional nao altera baseline ja estabelecido
    int n_antes = baseline_n;
    float v_extra = 200.0f;
    if (baseline_n < BASELINE_AMOSTRAS && v_extra > 100.0f) baseline_n++;
    check(baseline_n == n_antes, "Baseline nao e alterado apos estabelecido");
}

// 4. Deteccao de anomalia 
void test_deteccao_anomalia() {
    std::cout << "- Deteccao de anomalia\n";

    float baseline       = 145.0f;
    float variacao_severa = 20.0f;

    auto detecta = [&](float leitura) {
        return std::fabs(leitura - baseline) > variacao_severa;
    };

    check(!detecta(148.0f), "Leitura normal (148) nao dispara alarme");
    check(!detecta(145.0f), "Leitura identica ao baseline nao dispara alarme");
    check( detecta(180.0f), "Buraco (teto mais distante: 180) detectado");
    check( detecta(110.0f), "Saliencia (teto mais proximo: 110) detectada");
    check(!detecta(164.9f), "Leitura no limite (164.9, variacao=19.9) nao dispara");
    check( detecta(165.1f), "Leitura acima do limite (165.1, variacao=20.1) dispara");

    // Limiar configuravel
    variacao_severa = 50.0f;
    check(!detecta(180.0f), "Limiar aumentado para 50: leitura 180 nao dispara mais");
    check( detecta(200.0f), "Limiar aumentado para 50: leitura 200 ainda dispara");
}

// 5. Calculo de confianca
void test_confianca() {
    std::cout << "- Confianca\n";

    float zona_anterior = -999.0f;
    int   contagem      = 0;

    auto medir = [&](float pos) -> float {
        if (std::fabs(pos - zona_anterior) < 1.0f)
            contagem = std::min(contagem + 1, 10);
        else {
            zona_anterior = pos;
            contagem = 1;
        }
        return contagem / 10.0f;
    };

    float c1 = medir(0.0f);
    check(std::fabs(c1 - 0.1f) < 0.01f, "Primeira medicao na zona = 10%");

    for (int i = 0; i < 4; i++) medir(0.0f);
    check(std::fabs(contagem / 10.0f - 0.5f) < 0.01f, "5 medicoes na mesma zona = 50%");

    for (int i = 0; i < 10; i++) medir(0.0f);
    check(contagem == 10, "Contagem limitada a 10 (nao ultrapassa 100%)");

    float c_nova = medir(5.0f);
    check(std::fabs(c_nova - 0.1f) < 0.01f, "Mudanca de zona reinicia contagem para 10%");
}

// 6. Logica de troca de modo 
void test_modo_navegacao() {
    std::cout << "- Modo de navegacao\n";

    bool e_automatico = false;
    bool c_automatico = false;
    bool c_man        = false;

    // Ativar automatico
    c_automatico = true;
    if (c_automatico) { e_automatico = true;  c_automatico = false; }
    check(e_automatico  == true,  "c_automatico ativa e_automatico");
    check(c_automatico  == false, "c_automatico e resetado apos processamento");

    // Ativar manual
    c_man = true;
    if (c_man) { e_automatico = false; c_man = false; }
    check(e_automatico == false, "c_man desativa e_automatico");
    check(c_man        == false, "c_man e resetado apos processamento");

    // Comandos simultaneos: manual tem precedencia (ordem de processamento)
    c_automatico = true;
    c_man        = true;
    if (c_automatico) { e_automatico = true;  c_automatico = false; }
    if (c_man)        { e_automatico = false; c_man = false; }
    check(e_automatico == false, "Quando ambos os comandos chegam juntos, manual tem precedencia");
}

// 7. Clamp da saida do PID
void test_pid_clamp() {
    std::cout << "- PID clamp\n";

    auto clamp = [](float v, float lo, float hi) {
        return v < lo ? lo : (v > hi ? hi : v);
    };

    check(clamp( 150.0f, -100.0f, 100.0f) ==  100.0f, "Saida acima de 100 limitada a 100");
    check(clamp(-150.0f, -100.0f, 100.0f) == -100.0f, "Saida abaixo de -100 limitada a -100");
    check(clamp(  50.0f, -100.0f, 100.0f) ==   50.0f, "Saida dentro dos limites nao e alterada");
    check(clamp(   0.0f, -100.0f, 100.0f) ==    0.0f, "Saida zero nao e alterada");
    check(clamp( 100.0f, -100.0f, 100.0f) ==  100.0f, "Saida exatamente no limite superior e mantida");
    check(clamp(-100.0f, -100.0f, 100.0f) == -100.0f, "Saida exatamente no limite inferior e mantida");
}

// main
int main() {

    std::cout << " Testes do Sistema de Inspecao de Tuneis\n\n";

    test_encoder();             std::cout << "\n";
    test_filtro_media_movel();  std::cout << "\n";
    test_baseline();            std::cout << "\n";
    test_deteccao_anomalia();   std::cout << "\n";
    test_confianca();           std::cout << "\n";
    test_modo_navegacao();      std::cout << "\n";
    test_pid_clamp();           std::cout << "\n";

    std::cout << " Resultado: " << testes_passados << " OK  |  "
              << testes_falhos << " FALHOU\n\n";

    return testes_falhos > 0 ? 1 : 0;
}

File: src/test_sistema.py
#!/usr/bin/env python3

#Testes do modulo de simulacao: pool de anomalias, encoder, LIDAR, convergencia de
#velocidade e calculo de confianca.

import sys
import math
import random

testes_passados = 0
testes_falhos   = 0

def check(condicao, nome):
    global testes_passados, testes_falhos
    if condicao:
        print(f"[OK]     {nome}")
        testes_passados += 1
    else:
        print(f"[FALHOU] {nome}")
        testes_falhos += 1


# 1. Pool deterministico de anomalias 
def test_pool_deterministico():
    print("- Pool deterministico")

    def gerar_pool(n=200):
        return [(random.randint(200, 700),
                 random.choice(["buraco", "protuberancia"]),
                 random.randint(30, 70)) for _ in range(n)]

    random.seed(42)
    pool_a = gerar_pool()
    random.seed(42)
    pool_b = gerar_pool()

    check(pool_a == pool_b,
          "Mesma seed gera sequencia identica em ambas as interfaces")
    check(all(200 <= e[0] <= 700 for e in pool_a),
          "Todos os espacamentos estao no intervalo [200, 700]")
    check(all(e[1] in ("buraco", "protuberancia") for e in pool_a),
          "Todos os tipos sao validos")
    check(all(30 <= e[2] <= 70 for e in pool_a),
          "Todas as curvas estao no intervalo [30, 70]")

    # Diferentes seeds geram sequencias diferentes
    random.seed(1)
    pool_c = gerar_pool()
    check(pool_a != pool_c,
          "Seeds diferentes geram sequencias diferentes")


# 2. Encoder: alternancia a cada 100 pixels
def test_encoder():
    print("- Encoder")

    encoder_ant  = False
    transicoes   = 0
    for px in range(1, 501):
        enc = bool((px // 100) % 2)
        if enc != encoder_ant:
            transicoes += 1
            encoder_ant = enc

    check(transicoes == 5,
          f"500 pixels geram 5 transicoes de encoder (obteve {transicoes})")

    # Verifica alternancia binaria
    estados = [bool((px // 100) % 2) for px in range(0, 600, 100)]
    alternados = all(estados[i] != estados[i+1] for i in range(len(estados)-1))
    check(alternados,
          "Estados do encoder alternam corretamente a cada 100 pixels")


# 3. LIDAR: ruido gaussiano
def test_ruido_lidar():
    print("- LIDAR com ruido")

    random.seed(0)
    base    = 145.0
    n       = 2000
    amostras = [max(0, base + random.gauss(0, 2.0)) for _ in range(n)]

    media = sum(amostras) / n
    desvio = math.sqrt(sum((x - media)**2 for x in amostras) / n)

    check(abs(media - base) < 0.3,
          f"Media do ruido proxima do valor real (erro={media - base:.3f})")
    check(1.7 < desvio < 2.3,
          f"Desvio padrao proximo de 2.0 (obteve {desvio:.2f})")
    check(all(v >= 0 for v in amostras),
          "Nenhuma amostra e negativa (max(0, ...))")

    # Anomalia deve sobressair ao ruido
    variacao_buraco  = 40.0   # buraco grande
    variacao_ruido   = 2.0 * 3  # 3-sigma
    check(variacao_buraco > variacao_ruido,
          "Variacao de anomalia (40px) supera 3-sigma do ruido (6px)")


# 4. Convergencia de velocidade 
def test_convergencia_velocidade():
    print("- Convergencia de velocidade")

    delta = 1 / 30      # 30 fps
    fator = 15.0

    # De parado para velocidade normal (j_sp=50 -> alvo=200 px/s)
    vel   = 0.0
    alvo  = 200.0
    frames = 0
    while abs(vel - alvo) > 1.0 and frames < 1000:
        vel += (alvo - vel) * min(delta * fator, 1.0)
        frames += 1

    check(abs(vel - alvo) <= 1.0,
          f"Velocidade converge para o alvo em {frames} frames ({frames/30*1000:.0f} ms)")
    check(frames <= 10,
          f"Convergencia em menos de 10 frames (~333 ms) com fator {fator}")

    # Inspecao: alvo = (50 - 30) * 4 = 80 px/s
    vel   = 200.0
    alvo  = 80.0
    frames = 0
    while abs(vel - alvo) > 1.0 and frames < 1000:
        vel += (alvo - vel) * min(delta * fator, 1.0)
        frames += 1

    check(abs(vel - alvo) <= 1.0,
          f"Desaceleracao de inspecao converge em {frames} frames ({frames/30*1000:.0f} ms)")
    check(frames <= 10,
          "Desaceleracao em menos de 10 frames (~333 ms)")


# 5. Confianca online
def test_confianca():
    print("- Confianca online")

    zona_anterior = -999.0
    contagem      = 0

    def medir(pos):
        nonlocal zona_anterior, contagem
        if abs(pos - zona_anterior) < 1.0:
            contagem = min(contagem + 1, 10)
        else:
            zona_anterior = pos
            contagem = 1
        return contagem / 10.0

    c = medir(0.0)
    check(abs(c - 0.1) < 0.001, "Primeira medicao na zona = 10%")

    for _ in range(4):
        medir(0.0)
    check(abs(contagem / 10.0 - 0.5) < 0.001, "5 medicoes na zona = 50%")

    for _ in range(10):
        medir(0.0)
    check(contagem == 10, "Contagem limitada a 10 (nao ultrapassa 100%)")

    c_nova = medir(5.0)
    check(abs(c_nova - 0.1) < 0.001, "Mudanca de zona reinicia contagem para 10%")

    # Posicao proxima (< 1.0) deve manter zona
    c_prox = medir(5.3)
    check(contagem == 2, "Posicao dentro da tolerancia de 1m (5.3 vs 5.0) mantem zona atual")


# 6. Baseline e deteccao de anomalia
def test_deteccao_anomalia():
    print("- Baseline e deteccao de anomalia")

    BASELINE_N = 15
    baseline_soma = 0.0
    baseline_n    = 0
    baseline      = None
    valor_real    = 145.0
    variacao_severa = 20.0

    for _ in range(BASELINE_N):
        v = valor_real
        if baseline_n < BASELINE_N and v > 100.0:
            baseline_soma += v
            baseline_n += 1
            if baseline_n == BASELINE_N:
                baseline = baseline_soma / BASELINE_N

    check(baseline_n == BASELINE_N, "Baseline acumula exatamente 15 amostras")
    check(baseline is not None and abs(baseline - 145.0) < 0.01,
          "Baseline calculado corretamente (145.0)")

    detecta = lambda v: abs(v - baseline) > variacao_severa

    check(not detecta(148.0), "Leitura normal (148) nao dispara alarme")
    check(    detecta(180.0), "Buraco (180) detectado")
    check(    detecta(110.0), "Saliencia (110) detectada")

    variacao_severa = 50.0
    detecta2 = lambda v: abs(v - baseline) > variacao_severa
    check(not detecta2(180.0), "Limiar configuravel: 180 nao dispara com limiar=50")
    check(    detecta2(200.0), "Limiar configuravel: 200 ainda dispara com limiar=50")


#  main 
if __name__ == "__main__":

    print(" Testes da Simulacao (Python)\n")

    test_pool_deterministico(); print()
    test_encoder();             print()
    test_ruido_lidar();         print()
    test_convergencia_velocidade(); print()
    test_confianca();           print()
    test_deteccao_anomalia();   print()

    print(f" Resultado: {testes_passados} OK  |  {testes_falhos} FALHOU\n\n")

    sys.exit(1 if testes_falhos > 0 else 0)
